gooddata-flight-server/gooddata_flight_server/utils/methods_discovery.py

```diff
@@ -0,0 +1,58 @@
+#  (C) 2024 GoodData Corporation
+import importlib
+from functools import wraps
+from typing import Optional
+
+from gooddata_flight_server.server.base import (
+    FlightServerMethods,
+    FlightServerMethodsFactory,
+    ServerContext,
+)
+
+
+def flight_server_methods(fun: FlightServerMethodsFactory) -> FlightServerMethodsFactory:
+    """
+    Decorator that marks a function as the FlightServerMethodsFactory to be used.
+    Any module that should provide the methods factory should contain exactly one function decorated with this decorator.
+    """
+
+    @wraps(fun)
+    def wrapped(ctx: ServerContext) -> FlightServerMethods:
+        return fun(ctx)
+
+    wrapped.__is_flight_server_methods__ = True  # type: ignore
+    return wrapped
+
+
+def _is_flight_server_methods(obj: object) -> bool:
+    """
+    Checks whether the object is a FlightServerMethods.
+    """
+
+    return getattr(obj, "__is_flight_server_methods__", False)
+
+
+def get_methods_factory(module_name: str, root: Optional[str] = None) -> FlightServerMethodsFactory:
+    """
+    Get the method factory from the given module.
+    The module should contain exactly one method decorated with @flight_server_methods.
+
+    :param module_name: name of the module containing the methods abstract factory.
+    :param root: root package of the module: this is used to resolve relative module names (mainly useful in tests).
+    :return: a FlightServerMethodsFactory
+    """
+
+    module = importlib.import_module(module_name, package=root)
+
+    methods: list[FlightServerMethodsFactory] = [
+        member.__wrapped__  # unwrap the actual instance from the marker decorator to keep the type
+        for member in module.__dict__.values()
+        if _is_flight_server_methods(member)
+    ]
+
+    if len(methods) != 1:
+        raise ValueError(
+            f"Expected exactly one FlightServerMethodsFactory decorated with @flight_server_methods in {module_name}, got {len(methods)}"
```

Comment:
imho this error message should be expanded and be more 'beginner' friendly. also, the flight-server code should ensure that when a particular type of error happens when calling the discovered factory function, then the message is printed out 'nicely'; without stack-trace and other nasty stuff.

it may be better to have a dedicated exception (flight server may already have some that can be reused?) for these types of errors - to avoid scenario when ValueError flies due to genuine bug/error and our code then tosses away the stacktrace.

+ i suggest also cover (via error handling/messaging) the case when someone codes - by mistake and against the contract - factory function with less or more parameters (thus leading to errors).

+ perhaps the scenario when the decorated function returns nothing (again, error handling / messaging only).

---

our messaging in these cases needs to be.. you made a mistake coding the function. you should only annotate a function that has the right signature. correct your code, check out docs, blabla.